clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -4536,26 +4548,166 @@ void SYCLIntegrationFooter::emitSpecIDName(raw_ostream &O, const VarDecl *VD) {
   O << "";
 }
 
-bool SYCLIntegrationFooter::emit(raw_ostream &O) {
+template <typename BeforeFn, typename AfterFn>
+static void PrintNSHelper(BeforeFn Before, AfterFn After, raw_ostream &OS,
+                          const DeclContext *DC) {
+  if (DC->isTranslationUnit())
+    return;
+
+  const auto *CurDecl = cast<Decl>(DC);
+  // Ensure we are in the canonical version, so that we know we have the 'full'
+  // name of the thing.
+  CurDecl = CurDecl->getCanonicalDecl();
```

Comment:
Just double-checking my own understanding, this code is only called at the "end" of Sema, right? e.g., we had better not see additional redeclarations after this point?

Reply:
Correct.  This happens after delayed template instantiation.  This part of the code is essentially a 'pre-codegen' pass.